vm/src/obj/objstr.rs

```diff
@@ -1495,73 +1496,77 @@ impl PySliceableSequence for String {
     type Sliced = String;
 
     fn do_slice(&self, range: Range<usize>) -> Self::Sliced {
-        to_graphemes(self)
-            .get(range)
-            .map_or(String::default(), |c| c.join(""))
+        // an inefficient but simple implementation
+        // TODO: don't check whether the range contains each and every index
+        self.chars()
+            .enumerate()
+            .filter(|(idx, _)| range.contains(idx))
+            .map(|(_, ch)| ch)
+            .collect::<String>()
     }
 
     fn do_slice_reverse(&self, range: Range<usize>) -> Self::Sliced {
-        to_graphemes(self)
-            .get_mut(range)
-            .map_or(String::default(), |slice| {
-                slice.reverse();
-                slice.join("")
-            })
+        // an inefficient but simple implementation
+        // TODO: don't check whether the range contains each and every index
+        // try to remove temp String
+       let slice = self.chars()
+            .enumerate()
+            .filter(|(idx, _)| range.contains(idx))
+            .map(|(_, ch)| ch)
+            .collect::<String>();
+
+        slice.chars().rev().collect()
```

Comment:
For various reasons I couldn't do ```.rev()``` after ```.enumerate()``` so this was my workaround.  It can probably be done better

Reply:
I’d suggest something like this to use the `range` directly instead of enumerating and checking each index if it fits:

```rust
self.chars()
    .skip(range.start)
    .take(range.end - range.start)
    .collect::<String>()
    .chars()
    .rev()
    .collect();
```

Unfortunately I don’t think we can get rid of this collect, because `Skip` doesn’t implement `DoubleEndedIterator`.

Reply:
Alternatively, to get rid of the collected string, you could first count all the chars and use that information:

```rust
let count = self.chars().count();

self.chars()
    .rev()
    .skip(count - range.end)
    .take(range.end - range.start)
    .collect()
```

See [this in playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ce8152952e5868ae553eb0679845a438).

Reply:
Cool, thanks.  I'll work on this later today.